### PR TITLE
[FLINK-7632] [document] Overhaul on Cassandra connector doc

### PR DESCRIPTION
**Pull request by mcfongtw**
## What is the purpose of the change
Refactor Cassandra connector documentation by providing 
- in-depth information about Cassandra Pojo sink and Cassandra Tuple sink.
- meaningful examples to show distinction between streaming over Java Tuple and Pojo data types.

## Brief change log
- Refactor the current usage of on Cassandra Sink w/ more in-depth information.
- Provides examples for Pojo and Java Tuple data types

## Verifying this change
Have tested w/ local doc server (w/ docker setup)

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive):no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no

## Documentation
  - Does this pull request introduce a new feature? no

### CODE REVIEW

File: docs/dev/connectors/cassandra.md [hunk 1]
```diff
@@ -78,76 +96,195 @@ Note that that enabling this feature will have an adverse impact on latency.
 
 <p style="border-radius: 5px; padding: 5px" class="bg-danger"><b>Note</b>: The write-ahead log functionality is currently experimental. In many cases it is sufficent to use the connector without enabling it. Please report problems to the development mailing list.</p>
 
+### Checkpointing and Fault Tolerance
+With checkpointing enabled, Cassandra Sink guarantees at-least-once delivery of action requests to C* instance.
 
-#### Example
+<p style="border-radius: 5px; padding: 5px" class="bg-danger"><b>Note</b>:However, current Cassandra Sink implementation does not flush the pending mutations  before the checkpoint was triggered. Thus, some in-flight mutations might not be replayed when the job recovered. </p>
```
Comment (zentol) [inline docs/dev/connectors/cassandra.md]: theres a double space after mutations

File: docs/dev/connectors/cassandra.md [hunk 2]
```diff
@@ -78,76 +96,195 @@ Note that that enabling this feature will have an adverse impact on latency.
...
+
+More details on [checkpoints docs]({{ site.baseurl }}/dev/stream/state/checkpointing.html) and [fault tolerance guarantee docs]({{ site.baseurl }}/dev/connectors/guarantees.html)
+
+To enable fault tolerant guarantee, checkpointing of the topology needs to be enabled at the execution environment:
```
Comment (zentol) [inline docs/dev/connectors/cassandra.md]: imo this is a bit redundant, we are already linking to the checkpoint docs after all.

File: docs/dev/connectors/cassandra.md [hunk 3]
```diff
@@ -78,76 +96,195 @@ Note that that enabling this feature will have an adverse impact on latency.
...
+{% endhighlight %}
+</div>
+</div>
+
+### Cassandra Sink Example for Streaming Java Tuple Data Type
+While storing the result with Java Tuple data type to a Cassandra sink, it is required to set a CQL upsert statement (via setQuery('stmt')) to persist each record back to database. With the upsert query cached as `PreparedStatement`, Cassandra connector internally converts each Tuple elements as parameters to the statement.
+
+For details about `PreparedStatement` and `BoundStatement`, please visit [DataStax Java Driver manual](https://docs.datastax.com/en/developer/java-driver/2.1/manual/statements/prepared/)
+
+Please note that if the upsert query were not set, an `IllegalArgumentException` would be thrown with the following error message `Query must not be null or empty.`
```
Comment (zentol) [inline docs/dev/connectors/cassandra.md]: I would leave this out. Its one of those things that are easily out-dated, and don't provide immediate value when reading the docs for the first time. If a user stumbles upon this the error message should be self-explanatory; if it isn't we should change it accordingly.

File: docs/dev/connectors/cassandra.md [hunk 4]
```diff
@@ -78,76 +96,195 @@ Note that that enabling this feature will have an adverse impact on latency.
...
+CassandraSink.addSink(result)
+        .setQuery("INSERT INTO example.wordcount(word, count) values (?, ?);")
+        .setHost("127.0.0.1")
+        .build();
+{% endhighlight %}
+</div>
+
+<div data-lang="scala" markdown="1">
+{% highlight scala %}
+CassandraSink.addSink(input.javaStream)
```
Comment (zentol) [inline docs/dev/connectors/cassandra.md]: we generally try to keep the java&scala examples self-contained, even if it requires duplicated information.

File: docs/dev/connectors/cassandra.md [hunk 5]
```diff
@@ -78,76 +96,195 @@ Note that that enabling this feature will have an adverse impact on latency.
...
 </div>
 
-The Cassandra sinks support both tuples and POJO's that use DataStax annotations.
-Flink automatically detects which type of input is used.
 
-Example for such a Pojo:
+### Cassandra Sink Example for Streaming POJO Data Type
+An example of streaming a POJO data type and store the same POJO entity back to Cassandra. In addition, this POJO implementation needs to follow [DataStax Java Driver Manual](http://docs.datastax.com/en/developer/java-driver/2.1/manual/object_mapper/creating/) to annotate the class as Cassandra connector internally maps each field of this entity to an associated column of the desginated Table using `com.datastax.driver.mapping.Mapper` class of DataStax Java Driver.
+
+Please note that if the upsert query was set, an `IllegalArgumentException` would be thrown with the following error message `Specifying a query is not allowed when using a Pojo-Stream as input.`
```
Comment (zentol) [inline docs/dev/connectors/cassandra.md]: same as bove

File: docs/dev/connectors/cassandra.md [hunk 6]
```diff
@@ -78,76 +96,195 @@ Note that that enabling this feature will have an adverse impact on latency.
...
+
+For each CQL defined data type for columns, please refer to [CQL Documentation](https://docs.datastax.com/en/cql/3.1/cql/cql_reference/cql_data_types_c.html)
```
Comment (zentol) [inline docs/dev/connectors/cassandra.md]: the first part of the sentence doesn't quite make sense to me, the wording is a bit odd.

File: docs/dev/connectors/cassandra.md [hunk 7]
```diff
@@ -78,76 +96,195 @@ Note that that enabling this feature will have an adverse impact on latency.
...
+                    }
+                }
+            }
+        })
 
-	@Column(name = "id")
-	private long id;
-	@Column(name = "value")
-	private String value;
+        .keyBy("word")
```
Comment (zentol) [inline docs/dev/connectors/cassandra.md]: why is this example different than the previous one about tuples?

**Comment by zentol** [comment]
Nice improvement to the docs, had a few comments.

File: docs/dev/connectors/cassandra.md (hunk 7 above)
Comment (mcfongtw) [reply docs/dev/connectors/cassandra.md]: Thanks. Do you refer difference as for the string splitting or overloading keyBy methods? I will unify string splitting part. As for the keyBy, perhaps user could get to know that it is valid to use keyBy("field name") corresponding to an accessible field in the Pojo class.

File: docs/dev/connectors/cassandra.md (hunk 1 above)
Comment (mcfongtw) [reply docs/dev/connectors/cassandra.md]: Good eye. Thanks

File: docs/dev/connectors/cassandra.md (hunk 2 above)
Comment (mcfongtw) [reply docs/dev/connectors/cassandra.md]: Very true. Thanks.

File: docs/dev/connectors/cassandra.md (hunk 4 above)
Comment (mcfongtw) [reply docs/dev/connectors/cassandra.md]: Okay.

File: docs/dev/connectors/cassandra.md (hunk 6 above)
Comment (mcfongtw) [reply docs/dev/connectors/cassandra.md]: Will rephrase. Thank you.

**Comment by mcfongtw** [comment]
@zentol, Could you take another look? Thank you.

File: docs/dev/connectors/cassandra.md [hunk 8]
```diff
@@ -35,37 +43,47 @@ To use this connector, add the following dependency to your project:
...
+## Installing Apache Cassandra
+There are multiple ways to bring up a Cassandra instance on local machine:
+
+1. Follow the instructions from [Cassandra Getting Started page](http://cassandra.apache.org/doc/latest/getting_started/index.html).
+2. Launch a container running Cassandra from [Official Docker Repository](https://hub.docker.com/_/cassandra/)
 
-#### Installing Apache Cassandra
-Follow the instructions from the [Cassandra Getting Started page](http://wiki.apache.org/cassandra/GettingStarted).
+## Cassandra Sink
+Flink Cassandra connector currently supports Apache Cassandra as a data sink.
```
Comment (zentol) [inline docs/dev/connectors/cassandra.md]: This line is a bit weird, i Would just remove it.

File: docs/dev/connectors/cassandra.md [hunk 9]
```diff
@@ -35,37 +43,47 @@ To use this connector, add the following dependency to your project:
...
 
-#### Cassandra Sink
+### Configurations
 
 Flink's Cassandra sink are created by using the static CassandraSink.addSink(DataStream<IN> input) method.
```
Comment (zentol) [inline docs/dev/connectors/cassandra.md]: sink -> sinks; remove by (while we're at it)

File: docs/dev/connectors/cassandra.md [hunk 10]
```diff
@@ -35,37 +43,47 @@ To use this connector, add the following dependency to your project:
...
 
 The following configuration methods can be used:
 
-1. setQuery(String query)
-2. setHost(String host[, int port])
-3. setClusterBuilder(ClusterBuilder builder)
-4. enableWriteAheadLog([CheckpointCommitter committer])
-5. build()
+1. _setQuery(String query)_
+    * sets the upsert query that is executed for every record the sink receives.
```
Comment (zentol) [inline docs/dev/connectors/cassandra.md]: sets -> Sets

File: docs/dev/connectors/cassandra.md [hunk 11]
```diff
@@ -35,37 +43,47 @@ To use this connector, add the following dependency to your project:
...
+    * internally treated as CQL prepared statement, in which parameters could be shared or anonymous.
```
Comment (zentol) [inline docs/dev/connectors/cassandra.md]: -> The query is internally treated as a CQL ...

File: docs/dev/connectors/cassandra.md [hunk 12]
```diff
@@ -35,37 +43,47 @@ To use this connector, add the following dependency to your project:
...
+    * __DO__ set the upsert query for processing __Tuple__ data type
```
Comment (zentol) [inline docs/dev/connectors/cassandra.md]: missing period.

File: docs/dev/connectors/cassandra.md [hunk 13]
```diff
@@ -35,37 +43,47 @@ To use this connector, add the following dependency to your project:
...
+    * __DO NOT__ set the query for processing __POJO__ data type.
```
Comment (zentol) [inline docs/dev/connectors/cassandra.md]: type -> types

File: docs/dev/connectors/cassandra.md [hunk 14]
```diff
@@ -35,37 +43,47 @@ To use this connector, add the following dependency to your project:
...
+2. _setClusterBuilder()_
+    * sets the cluster builder that is used to configure the connection to cassandra with more sophisticated settings such as consistency level, retry policy and etc.
```
Comment (zentol) [inline docs/dev/connectors/cassandra.md]: sets -> Sets

File: docs/dev/connectors/cassandra.md [hunk 15]
```diff
@@ -35,37 +43,47 @@ To use this connector, add the following dependency to your project:
...
+3. _setHost(String host[, int port])_
+    * simple version of setClusterBuilder() with host/port information to connect to Cassandra instances
+4. _enableWriteAheadLog([CheckpointCommitter committer])_
+    * an __optional__ setting
+    * allows exactly-once processing for non-deterministic algorithms.
+5. _build()_
+    * finalizes the configuration and constructs the CassandraSink instance.
```
Comment (zentol) [inline docs/dev/connectors/cassandra.md]: finalizes -> Finalizes

File: docs/dev/connectors/cassandra.md [hunk 16]
```diff
@@ -78,77 +96,189 @@ Note that that enabling this feature will have an adverse impact on latency.
...
+    word text,
+    count bigint,
+    PRIMARY KEY(word)
+    );
+{% endhighlight %}
+</div>
+</div>
+
+### Cassandra Sink Example for Streaming Tuple Data Type
+While storing the result with Java/Scala Tuple data type to a Cassandra sink, it is required to set a CQL upsert statement (via setQuery('stmt')) to persist each record back to database. With the upsert query cached as `PreparedStatement`, Cassandra connector internally converts each Tuple elements as parameters to the statement.
```
Comment (zentol) [inline docs/dev/connectors/cassandra.md]: -> back to **the** database.

I would reword the second part of second sentence to something like ", each Tuple element is converted to parameters of the statement."

File: docs/dev/connectors/cassandra.md [hunk 17]
```diff
@@ -78,77 +96,189 @@ Note that that enabling this feature will have an adverse impact on latency.
...
+                    if (!word.isEmpty()) {
+                        out.collect(new Tuple2<String, Long>(word, 1L));
+                    }
+                }
+            }
+        })
+        .keyBy(0)
+        .timeWindow(Time.seconds(5))
+        .sum(1)
+        ;
```
Comment (zentol) [inline docs/dev/connectors/cassandra.md]: move to previous line

File: docs/dev/connectors/cassandra.md [hunk 18]
```diff
@@ -78,77 +96,189 @@ Note that that enabling this feature will have an adverse impact on latency.
...
 </div>
+
 </div>
 
-The Cassandra sinks support both tuples and POJO's that use DataStax annotations.
-Flink automatically detects which type of input is used.
 
-Example for such a Pojo:
+### Cassandra Sink Example for Streaming POJO Data Type
+An example of streaming a POJO data type and store the same POJO entity back to Cassandra. In addition, this POJO implementation needs to follow [DataStax Java Driver Manual](http://docs.datastax.com/en/developer/java-driver/2.1/manual/object_mapper/creating/) to annotate the class as Cassandra connector internally maps each field of this entity to an associated column of the desginated Table using `com.datastax.driver.mapping.Mapper` class of DataStax Java Driver.
```
Comment (zentol) [inline docs/dev/connectors/cassandra.md]: -> to annotate the class **as each field of this entity is mapped to an associated column of the designated table using the DataStax Java Driver `com.datastax.driver.mapping.Mapper` class.**

File: docs/dev/connectors/cassandra.md (hunk 7 above)
Comment (zentol) [reply docs/dev/connectors/cassandra.md]: ah nevermind, i didn't saw that the Tuple example does a keyBy too.

File: docs/dev/connectors/cassandra.md (hunk 11 above)
Comment (mcfongtw) [reply docs/dev/connectors/cassandra.md]: It is true that we don't need to get to that level of details. I could rephrase to what you suggested.

**Comment by mcfongtw** [comment]
Hi, @zentol , could you take another look? Thank you.

**Comment by zentol** [comment]
Looks really god now, merging.
